src/main/scala/com/advancedtelematic/campaigner/db/ExtractUpdatesFromCampaignsAndInsert.scala

```diff
@@ -0,0 +1,49 @@
+package com.advancedtelematic.campaigner.db
+
+import java.time.Instant
+
+import akka.actor.ActorSystem
+import akka.stream.Materializer
+import com.advancedtelematic.campaigner.data.DataType._
+import com.advancedtelematic.campaigner.db.Schema.{campaigns, updates}
+import com.advancedtelematic.libats.messaging_datatype.DataType.UpdateId
+import com.advancedtelematic.libats.slick.db.SlickExtensions._
+import com.advancedtelematic.libats.slick.db.SlickUUIDKey._
+import org.slf4j.LoggerFactory
+import slick.jdbc.MySQLProfile.api._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+
+class ExtractUpdatesFromCampaignsAndInsert(implicit
+                                           val db: Database,
+                                           val mat: Materializer,
+                                           val system: ActorSystem,
+                                           val ec: ExecutionContext
+                                   ) {
+
+  private val _log = LoggerFactory.getLogger(this.getClass)
+
+  private[db] def updateFromCampaign(c: Campaign): Update = {
+    val u = Update(UpdateId.generate(), UpdateSource(c.updateId.uuid.toString, UpdateType.multi_target), c.namespace, c.name, None, Instant.now, Instant.now)
+    _log.info("Extracted update {} from campaign {}", u, c, "")
+    u
+  }
+
+  private[db] def createOneUpdateRecordForEachCampaign(): Future[Option[Int]] =
+    db.run(campaigns.result)
+      .map { cs => cs.map(updateFromCampaign) }
+      .map { us => us.groupBy(_.source).map(_._2.head) } // filter UpdateSource duplicates
+      .map { us => us.groupBy(u => (u.namespace, u.source.id)).map(_._2.head) } // filter unique key (namespace, updateId) duplicates (shouldn't be necessary after the previous filter?)
+      .flatMap { us => db.run(updates ++= us) }
+
+  def run() = {
```

Comment:
So if I am reading this correctly, run could be:

```scala
  def run() = db.run {
    campaigns.distinctOn(_.update).result
      .map { cs => cs.map(c => c.id -> updateFromCampaign(c)).toMap }
      .flatMap { us => (updates ++= us.values).map(_ => us) }
      .flatMap { us =>
        DBIO.sequence(
          us.foldLeft(List.empty[DBIO[Int]]) { case (acc, (campaignId, update)) =>
            campaigns.filter(_.id === campaignId).map(_.update).update(update.uuid) :: acc
          }
        )
      }
      .transactionally
      .map(_ => Done)
  }
```

Reply:
Awesome! I spent so many hours trying to come up with a "all in one go" query like that :D Indeed it looks like what we need, I tested it.
I just changed the first line with a workaround because there seems to be a bug with `distinctOn` and it was giving me duplicates, have a look at the comment in the code.

Reply:
that is exactly why I don't like Slick

Reply:
That is not a very good reason not to like slick... you could just write this in sql as well. https://github.com/advancedtelematic/ota-tuf/commit/6fd15d0#diff-cb2654b1aa07bb1bff4cf75ba0f3e614R23 

But anyway, yeah this is shitty. You can leave the workaround there because data is at the moment very small but this could easily blow up, and you would not be able to do something like this with db.stream.